### PR TITLE
OC-771 - Displaying all cross links

### PR DESCRIPTION
**Pull request by finlay-jisc**
The purpose of this PR was to give users a way of viewing all the crosslinks associated with a publication, further to the related publications sidebar which is capped at showing 5 links.

---

### Acceptance Criteria:

- Whilst viewing a publication page and there are more than 5 suggested links, a button to “Show All” is present at the bottom of the “Related publications” area, opening the “Related Content” modal. 
- The “Related Content” modal shows a list of each publication that has been suggested for linking with this publication, including title, publication date and first author.
    - A search control is present at the top of the modal to search user suggested links by publication title
    - Above the publication list, the text “Sort by” is present, with options available to sort the list by “Most recent” and “Most relevant”, presented as two radio buttons (not present in mockup).
        - By default, the list is sorted by recency, with the newest links first
    - To the right of the sort options, a checkbox is present if the user is logged in, reading “View my suggestions” which filters to links that the author has suggested.
    - The list of publications is paginated, with X publications displayed per page.
        - The same style of pagination control is used as is shown on the main publication search page.
        - A delete button is present against any suggestions the user has made.
            - Clicking this delete button will remove the suggested link.
        - An “opens in a new tab” style button is present to view each publication
    - A close button is present below the pagination options at the bottom of the modal (not in mockup)
---

### Checklist:

- [x] Local manual testing conducted
- [x] Automated tests added
- [ ] Documentation updated

---

### Tests:

UI
Screenshot 2024-05-09 150805

API
Screenshot 2024-05-09 153541

E2E
Screenshot 2024-05-10 113839

---

### Screenshots:
Screenshot 2024-05-10 115409

### CODE REVIEW

File: api/src/components/crosslink/controller.ts [hunk 1]
```diff
@@ -152,10 +157,37 @@ export const getPublicationCrosslinks = async (
             return response.json(404, { message: 'Publication not found.' });
         }
 
-        const crosslinks = await crosslinkService.getPublicationCrosslinks(
-            event.pathParameters.publicationId,
-            event.queryStringParameters.order || undefined
-        );
+        const ownLinks = event.queryStringParameters.own === 'true';
+
+        if (ownLinks && !event.user) {
```
Comment (adwearing-jisc) [inline api/src/components/crosslink/controller.ts:162]: For these two combinations you're checking for, are they prevented in the UI? Therefore these checks are just incase someone manipulates the API request?

File: api/src/components/crosslink/controller.ts [hunk 2]
```diff
@@ -152,10 +157,37 @@ export const getPublicationCrosslinks = async (
...
+        if (
+            event.queryStringParameters.order === 'mix' &&
+            (event.queryStringParameters.limit || event.queryStringParameters.offset)
+        ) {
+            return response.json(400, {
+                message: 'Pagination cannot be used with "mix" sort order.'
+            });
+        }
+
+        const search = event.queryStringParameters?.search
```
Comment (adwearing-jisc) [inline api/src/components/crosslink/controller.ts:177]: Curious why this one needs the Optional Chaining but all the others above don't

File: api/src/components/crosslink/service.ts [hunk 3]
```diff
@@ -84,8 +86,36 @@ export const exists = async (crosslinkId: string) => {
...
+        select: {
+            votes: {
+                select: {
+                    vote: true
+                }
+            }
+        }
+    });
+
+    const score = crosslink?.votes
```
Comment (adwearing-jisc) [inline api/src/components/crosslink/service.ts:103]: Handling crosslink being null?

File: api/src/components/crosslink/controller.ts (hunk 1 above)
Comment (finlay-jisc) [reply api/src/components/crosslink/controller.ts:162]: For now, pretty much yes. Generally on octopus I've been putting this kind of check in at both the API & UI level. The API is starting to get a bit more standalone now so I'm mainly trying to ensure that no validation slip ups make it through. Even though something like this is unlikely to be used without the UI, I'm just following that model everywhere to be safe.

File: api/src/components/crosslink/controller.ts (hunk 2 above)
Comment (finlay-jisc) [reply api/src/components/crosslink/controller.ts:177]: Not sure why it was there, it's not needed!

File: api/src/components/crosslink/service.ts (hunk 3 above)
Comment (finlay-jisc) [reply api/src/components/crosslink/service.ts]: Yeah, that's missing. Just added a bit to return false if it is null.

**Review by adwearing-jisc — APPROVED** [review]

**Pull request merged**